test/ably/realtime/realtimeauth_test.py

```diff
@@ -323,3 +324,79 @@ async def auth_callback(_):
         assert ably.auth.token_details is not original_token_details
 
         await ably.close()
+
+    # RTN14b
+    async def test_renew_token_single_attempt(self):
+        rest = await TestApp.get_ably_rest()
+
+        async def callback(params):
+            token_details = await rest.auth.request_token(token_params=params)
+            return token_details.token
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        msg = {
+            "action": ProtocolMessageAction.ERROR,
+            "error": {
+                "code": 40142,
+                "statusCode": 401
+            }
+        }
+
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        original_token_details = ably.auth.token_details
+        await ably.connection.connection_manager.transport.on_protocol_message(msg)
+        assert ably.auth.token_details is not original_token_details
+        await ably.close()
+
+    # RTN14b
+    async def test_renew_token_connection_attempt_fails(self):
+        rest = await TestApp.get_ably_rest()
+
+        async def callback(params):
+            token_details = await rest.auth.request_token(token_params=params)
+            return token_details.token
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        msg = {
+            "action": ProtocolMessageAction.ERROR,
+            "error": {
+                "code": 40142,
+                "statusCode": 401
+            }
+        }
+
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        ably.connection.connection_manager.enact_state_change(ConnectionState.DISCONNECTED,
+                                                              AblyException("token error", 401, 40143))
+        await ably.connection.connection_manager.transport.on_protocol_message(msg)
+        state_change = await ably.connection.once_async(ConnectionState.DISCONNECTED)
+        assert ably.connection.error_reason == state_change.reason
+        assert state_change.reason.code == 40143
+        assert state_change.reason.status_code == 401
+        await ably.close()
+
+    # RSA4a
+    async def test_renew_token_no_renew_means_provided(self):
+        rest = await TestApp.get_ably_rest()
+
+        async def callback(params):
+            token_details = await rest.auth.request_token(token_params=params)
+            return token_details.token
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        msg = {
```

Comment:
this test also doesn't make sense because this error will never be received from ably, the best way to test this is to initialise the library with just a token (or token_details) and simulate receipt of a DISCONNECTED message with a token error